### ISSUE TITLE
Frequently crash of libgowhatsapp.dll

### ISSUE THREAD

**Issue by pkar70** [hoehermann/purple-gowhatsapp#56]
I added this protocol, QR code... it starts to add contacts. Crashes. Reopen pidgin, it works. I started to move buddies to sort buddies, after some time it crashes. Then I cannot reopen pidgin, every time it crashes. I disable protocol, open pidgin, wait some minutes, try to enable protocol - crash.
After about week, I tried again. Open, get messages, and several minutes later, crash.

Event log says:
Faulting application name: pidgin.exe, version: 2.13.0.0, time stamp: 0x00000000
Faulting module name: libgowhatsapp.dll, version: 0.0.0.0, time stamp: 0x5ea00288
Exception code: 0xc0000005
Fault offset: 0x00001b07
Faulting application path: C:\Program Files (x86)\Pidgin\pidgin.exe
Faulting module path: C:\Program Files (x86)\Pidgin\plugins\libgowhatsapp.dll

**Comment by EionRobb** [comment]
Can you include the backtrace from the %appdata%\.purple\Pidgin.rpt file

**Comment by pkar70** [comment]
Yes, and no :)
Yes, because I found it.
No, because last entry is dated May 6, 2020, and is from another module (when whatsapp was not active):
C:\Program Files (x86)\Pidgin\pidgin.exe caused an Access Violation at location 625a21d2 in module C:\Program Files (x86)\Pidgin\plugins\ssl-nss.dll Reading from location 00000004.

**Comment by pkar70** [comment]
Is there any way to get BuddyList:Help:Debug window content redirected to file? Not as "save" from this window, as this window disappears on crash :)

**Comment by pkar70** [comment]
From this debug window, while pidgin works:
(13:34:56) gowhatsapp: 3AFC2AD0 recieved message type 5 at 0 id (null) remote 48500892405@s.whatsapp.net sender (null) (fromMe 0, system 1): available
(but there is no new messages shown in conversation window)

**Comment by hoehermann** [comment]
@pkar70 Sorry this happens. Seems to be similar to #53.

> recieved message type 5

This is a "presence" type message. It only updates the "last seen" property of the buddy in the list.

> Is there any way to get BuddyList:Help:Debug window content redirected to file?

Close pidgin. Use the explorer to navigate to your pidgin installation (where `pidgin.exe` resides). Hold shift and right-click. Select "open PowerShell / command line here". Enter `.\pidgin.exe -d > pidgin.log`. Use pidgin until it crashes. Remove all sensitive information from `pidgin.log` and upload it here.

**Comment by leolca** [comment]
It is making pidgin crashes.

here is the end of the log result:

```
(09:43:28) gowhatsapp: 0x5570cd4c6ce0 recieved message type 1 at 1546860179 id XXXXXXXXXXXXXXXXXXXXXXXXXXXX EF3A remote 123456789@s.whatsapp.net sender  (fromMe 1, system 0): 👍
(09:43:28) g_log: purple_conversation_get_im_data: assertion 'conv != NULL' failed
(09:43:28) GLib-GObject: value "-864970583" of type 'gint' is invalid or out of range for property 'weight' of type 'gint'
(09:43:28) GLib-GObject: value "-864970583" of type 'gint' is invalid or out of range for property 'weight' of type 'gint'
```

**Comment by hoehermann** [comment]
@leolca Thank you for the hint. Unfortunately, I cannot reproduce this behaviour. In Wine, the emoji is not displayed, but it does not crash the application.

@EionRobb: Do you happen to know whether there are any known problems regarding extended use of unicode characters like emojis in Pidgin and/or gtk+ on Windows?

**Comment by EionRobb** [comment]
@hoehermann yes, pidgin on windows can't display them without an emoji smiley theme, but emoji smiley themes don't work with unicode on the current version 2.13 (was broken in 2.12 and fixed in 2.14) - none of them cause crashes though, just tofu/squares instead of characters

**Comment by hoehermann** [comment]
@pkar70 Is this issue resolved in version 0.3.2? Binaries for Ubuntu 18.04 and Windows are available at https://buildbot.hehoe.de/purple-gowhatsapp/builds/.

**Comment by pkar70** [comment]
Seems like on my setup works only libgowhatsapp.dll which has 16 MB, not 8 MB. So I cannot test it.
When I have 8 MB version of dll, in Accounts menu I have entry for whatsapp ("..@s.whatsapp.net (unknown)").
I'm guessing that 16 MB has static linking, and 8 MB is dynamic linking, with some libs I doesn't have.


Besides, I got crash only from time to time (but when it occurs, many re-start of pidgin are required to make it work; seems like starting pidgin with whatsapp protocol disabled, and enabling it from menu, somehow helps (but not always).

**Comment by hoehermann** [comment]
> I'm guessing that 16 MB has static linking, and 8 MB is dynamic linking, with some libs I doesn't have.

That is unfortunate. I was not aware this could be an issue. Can you use https://dependencywalker.com/ and tell me which libraries I need to link statically or supply explicitly? Might be GTK.

**Comment by pkar70** [comment]
It hangs on 16 MB version, and on 8 MB version it says that several are missing. Some are in pidgin.exe folder, not in plugins, these are probably accessible anyway.
libintl-8.dll seems like not present in my installation.

**Comment by hoehermann** [comment]
libintl-8.dll seems to be part of GTK. I downloaded it from [here](https://sourceforge.net/projects/gtk-mingw/files/gtk%2B2/gtk%2B-2.24.11-1-bundle.7z/download). As far as I know, one cannot link statically against GTK.

**Comment by EionRobb** [comment]
@hoehermann the version of GTK that winpidgin uses has intl.dll not libintl-8.dll

**Comment by hoehermann** [comment]
Ah, I see. https://buildbot.hehoe.de/purple-gowhatsapp/builds/libgowhatsapp_0.3.2~gite6b2bd9+gowhatsapp~git9b4bc38_win32.dll does not depend on libintl-8.dll. At least not directly. If you happen to have a rogue libglib-2.0-0.dll in your system which does, there is nothing I can do about it. @pkar70 Does this build work for you?

**Comment by pkar70** [comment]
I'm using dll from link in issue#50.

Dependency walker shows libintl-8.dll as 'first-step' dependency from ibgowhatsapp_0.3.2~gite6b2bd9+gowhatsapp~git9b4bc38_win32.dll.
![image](https://user-images.githubusercontent.com/23451507/88489080-5e26d700-cf92-11ea-9e2e-c626881cdb9d.png)

**Comment by EionRobb** [comment]
> Ah, I see. https://buildbot.hehoe.de/purple-gowhatsapp/builds/libgowhatsapp_0.3.2~gite6b2bd9+gowhatsapp~git9b4bc38_win32.dll does not depend on libintl-8.dll. At least not directly. If you happen to have a rogue libglib-2.0-0.dll in your system which does, there is nothing I can do about it. @pkar70 Does this build work for you?

This one is good, no intl dependency.  Did you add it in statically?
![image](https://user-images.githubusercontent.com/1063865/88489570-0e272f00-cfe9-11ea-81a7-4ebf796301f5.png)

**Comment by hoehermann** [comment]
@EionRobb 

> Did you add it in statically?

No. I am building against the old GTK+ 2.14.7 now. Also, I removed the `ENABLE_NLS` define. I am not sure which of these two measures were effective.

@pkar70 Your screenshot shows the dependencies of version 0.3.2~git8b37120. I am referring to version [0.3.2~gite6b2bd9](https://buildbot.hehoe.de/purple-gowhatsapp/builds/libgowhatsapp_0.3.2~gite6b2bd9+gowhatsapp~git9b4bc38_win32.dll). I did not change the version number while investigating the issue. Sorry for the confusion.

**Comment by EionRobb** [comment]
> No. I am building against the old GTK+ 2.14.7 now. Also, I removed the `ENABLE_NLS` define. I am not sure which of these two measures were effective.

It'll be the ENABLE_NLS... libintl was only being used for one function: libintl_gettext()
![image](https://user-images.githubusercontent.com/1063865/88489908-bc33d880-cfeb-11ea-8776-4e8df4a2a41c.png)

**Comment by hoehermann** [comment]
> It'll be the ENABLE_NLS

Indeed, you are correct.

However, there are other functions not available in GTK+ 2.14.7, e.g. `g_get_monotonic_time` and `g_slist_free_full` which are used by the `http` module which was taken from purple3. I want to keep building against the old version so I know of the problem.

**Comment by pkar70** [comment]
0.3.2~gite6b2bd9 works (so, it (a) doesn't require anything that I don't have; and (b) connects to whatsapp and allows to write/read messages).
